scalding-core/src/main/scala/com/twitter/scalding/ExecutionUtil.scala

```diff
@@ -0,0 +1,122 @@
+package com.twitter.scalding
+
+import scala.collection.mutable
+
+import com.twitter.algebird.Monoid
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success }
+
+object ExecutionUtil {
+  private[ExecutionUtil] class AsyncSemaphore(initialPermits: Int = 0) {
+    private[this] val waiters = new mutable.Queue[() => Unit]
+    private[this] var availablePermits = initialPermits
+
+    private[ExecutionUtil] class SemaphorePermit {
+      def release() =
+        AsyncSemaphore.this.synchronized {
+          availablePermits += 1
+          if (availablePermits > 0 && waiters.nonEmpty) {
+            availablePermits -= 1
+            waiters.dequeue()()
+          }
+        }
+    }
+
+    def acquire(): Future[SemaphorePermit] = {
+      val promise = Promise[SemaphorePermit]()
+
+      def setAcquired(): Unit =
+        promise.success(new SemaphorePermit)
+
+      synchronized {
+        if (availablePermits > 0) {
+          availablePermits -= 1
+          setAcquired()
+        } else {
+          waiters.enqueue(setAcquired)
+        }
+      }
+
+      promise.future
+    }
+  }
+
+  /**
+   * Run a sequence of executions but only permitting parallelism amount to run at the
+   * same time.
+   *
+   * @param executions List of executions to run
+   * @param parallelism Number to run in parallel
+   * @return Execution Seq
+   */
+  def withParallelism[T](executions: Seq[Execution[T]], parallelism: Int): Execution[Seq[T]] = {
```

Comment:
I don't see why not to put this in `object Execution`. Seems almost as general as sequence.